core/invoker/src/main/scala/whisk/core/containerpool/ContainerProxy.scala

```diff
@@ -0,0 +1,388 @@
+                            case _                               => ActivationResponse.whiskError(t.getMessage)
+                        }
+                        val activation = ContainerProxy.constructWhiskActivation(job, Interval.zero, response)
+                        sendActiveAck(transid, activation)
+                        storeActivation(transid, activation)
+                }
+                .flatMap {
+                    container =>
+                        run(container, job)
+                            .map(_ => WarmedData(container, job.msg.user.namespace, job.action, Instant.now))
+                }.pipeTo(self)
+
+            goto(Running)
+    }
+
+    when(Starting) {
+        // container was successfully obtained
+        case Event(data: PreWarmedData, _) =>
+            context.parent ! NeedWork(data)
+            goto(Started) using data
+
+        // container creation failed
+        case Event(_: FailureMessage, _) =>
+            context.parent ! ContainerRemoved
+            stop()
+
+        case _ => delay
+    }
+
+    when(Started) {
+        case Event(job: Run, data: PreWarmedData) =>
+            implicit val transid = job.msg.transid
+            run(data.container, job)
+                .map(_ => WarmedData(data.container, job.msg.user.namespace, job.action, Instant.now))
+                .pipeTo(self)
+
+            goto(Running)
+
+        case Event(Remove, data: PreWarmedData) => destroyContainer(data.container)
+    }
+
+    when(Running) {
+        // Intermediate state, we were able to start a container
+        // and we keep it in case we need to destroy it.
+        case Event(data: PreWarmedData, _) => stay using data
+
+        // Run was successful
+        case Event(data: WarmedData, _) =>
+            context.parent ! NeedWork(data)
+            goto(Ready) using data
+
+        // Failed after /init (the first run failed)
+        case Event(_: FailureMessage, data: PreWarmedData) => destroyContainer(data.container)
+
+        // Failed for a subsequent /run
+        case Event(_: FailureMessage, data: WarmedData)    => destroyContainer(data.container)
+
+        // Failed at getting a container for a cold-start run
+        case Event(_: FailureMessage, _) =>
+            context.parent ! ContainerRemoved
+            stop()
+
+        case _ => delay
+    }
+
+    when(Ready, stateTimeout = pauseGrace) {
+        case Event(job: Run, data: WarmedData) =>
+            implicit val transid = job.msg.transid
+            run(data.container, job)
+                .map(_ => WarmedData(data.container, job.msg.user.namespace, job.action, Instant.now))
+                .pipeTo(self)
+
+            goto(Running)
+
+        // pause grace timed out
+        case Event(StateTimeout, data: WarmedData) =>
+            data.container.halt()(TransactionId.invokerNanny).map(_ => ContainerPaused).pipeTo(self)
```

Comment:
fwiw finding a bit of mental friction with `halt` (which i think of as no only ceasing the CPU but also powering off a traditional system).